positional_encodings/positional_encodings.py

```diff
@@ -76,7 +76,40 @@ def forward(self, tensor):
         emb[:, :, : self.channels] = emb_x
         emb[:, :, self.channels : 2 * self.channels] = emb_y
 
-        return emb[None, :, :, :orig_ch].repeat(batch_size, 1, 1, 1)
+        return emb[None, :, :, :orig_ch]
+
+class FixedPositionalEncoding2D(nn.Module):
+    def __init__(self, shape, channels):
+        super(FixedPositionalEncoding2D, self).__init__()
+        self.pos_encoder = SinglePositionalEncoding2D(channels)
+        self.pos_encoding = self.pos_encoder(torch.ones(1, shape[0], shape[1], channels))
+        self.device = self.pos_encoding.device
+        self.repeated_pos_encoding = self.pos_encoding.repeat(32, 1, 1, 1)
+
+    def forward(self, tensor):
+        # Ensure encoding on same device
+        if self.device is not tensor.device:
+            self.pos_encoding = self.pos_encoding.to(tensor.device)
+            self.device = self.pos_encoding.device
+        # Match batch size
+        if self.repeated_pos_encoding.shape[0] != tensor.shape[0]:
+            self.repeated_pos_encoding = self.pos_encoding.repeat(tensor.shape[0], 1, 1, 1)
+        return self.repeated_pos_encoding
+
+class PositionalEncoding2D(nn.Module):
```

Comment:
do we need this class here? all it is doing is just setting up the positional encoding class from before and then just calling its forward method in this forward method.

My suggestion is to get rid of this class, and to leave the name of the other one, while just having the new one called `Fixed`...

Reply:
I think it would even be cleaner to merge the FixedPositonalBlaBla with the original PositinalBlaBla and have a key argument of shape. And if you provide shape, you will be using the fixed route and if you leave the shape you will be using the original route. So just 1 class instead of 3.